include/cell/copy/shared_to_register.hpp

```diff
@@ -221,26 +310,116 @@ struct SharedToRegLoader<Reg_, WarpLayout_, kMode, CopyInst::LoadMat> {
-    DEVICE void operator()(const Reg& src, Shared& dst) {}
-};
+template <typename Reg_, typename WarpLayout_>
+struct RegToSharedStorer<Reg_, WarpLayout_, RegLayout::WMMA_m16n16k16,
+                         CopyInst::LoadS32> {
+    using Reg = Reg_;
+    using WarpLayout = WarpLayout_;
+    using DType = typename Reg::DType;
 
-template <typename Reg, typename Shared>
-DEVICE void copy_tile_r2s(const Reg& src, Shared& dst) {
-    using Copy = RegToShared<Reg, Shared, InstShape<16, 16, 16>,
-                             RegLayout::TileWMMA, CopyInst::LoadS32>;
-    Copy copy;
-    copy(src, dst);
-}
+    template <typename Shared>
+    DEVICE void operator()(const Reg& src, Shared& dst) {
+        static_assert(std::is_same<typename Shared::DType, DType>::value,
+                      "The data type of Shared and Reg must be the same.");
+        static_assert(
+            (Reg::kNumel * 32 * tl::get_numel<WarpLayout>) == Shared::kNumel,
+            "The number of elements in Reg and Shared must be the same.");
+
+        const bool row_major = Shared::kIsRowMajor;  // get the layout of shared
+        const DType* src_ptr = src.data();    // pointer for register tile
+        DType* dst_ptr = dst.mutable_data();  // pointer for shared memory tile
+
+        int row_exec =
+            row_exec_count<DType, Shared::kRows, tl::num_rows<WarpLayout>,
+                           WarpReuse::Cont>();
+        int col_exec =
+            row_exec_count<DType, Shared::kCols, tl::num_cols<WarpLayout>,
+                           WarpReuse::Cont>();
+
+        // 1. advance the pointer to input data to the current warp according to
+        // warp reuse mode. During the store process, threads do not write to
+        // the same shared memory location, thus the warp reuse mode is set to
+        // `Cont`.
+        int offset = get_warp_offset<WarpReuse::Cont, Shared, WarpLayout>();
+        dst_ptr += offset;
+
+        // FIXME(haruhi): Try to find a better way to organize these
+        // hardware-dependent magic numbers.
+        // The number 2 is interpreted as follows: 8x8 block is a fundamental
+        // unit of a TCU instruction. Regardless of the output precision, a
+        // single execution of WMMA stores 2 elements in each thread's local
+        // register.
+        static constexpr int stride = 2;
+        static constexpr int rstride =
+            row_major ? tl::num_rows<ThreadWmma> * Shared::kRowStride
+                      : tl::num_rows<ThreadWmma>;
+        static constexpr int cstride =
+            row_major ? stride * tl::num_cols<ThreadWmma>
+                      : stride * tl::num_cols<ThreadWmma> * Shared::kColStride;
+
+        // TODO: Improve this is a naive implementation that does not use
+        // vectorized access in future. To use vectorized access, we need to
+        // know the return type of WMMA.
+        auto store_base_tile = [&](const DType* src, DType* dst) {
```

Comment:
Can we implemente it as a private function?

Reply:
Completed. Also, I addressed two other places that had similar issues.